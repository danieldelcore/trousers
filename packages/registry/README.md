# @trousers/registry

`@trousers/registry` is responsible for storing the state of the styles have been mounted, mounting them to the head of the document and also providing a way to clear the styles.

## API

### `registry()`

Is the factory function which exposes the registry API.

**Arguments:**

-   `parentElement`: HTMLElement
-   `attributeId`: string
-   `options`?: Object
    -   `forceNewNode`: boolean
    -   `appendBefore`: HTMLElement

**Returns:**

-   `register`: Function
-   `clear`: Function
-   `has`: Function

### `registry.register(id, styles, isGlobal)`

Registers provided styles. Please note that this method is significantly faster in prodmode because it uses [insertRule](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/insertRule). However the drawback is that you cannot augment the styles once they're mounted.

**Arguments:**

-   `id`: string - unique identifier for the provided style
-   `styles`: string - style string
-   `isGlobal`?: boolean - whether the styles should be treated as global or not

**Example:**

```jsx
import registry from '@trousers/registry';

const clientRegistry = registry(
    document.createElement('head'),
    'i-love-trousers',
);
clientRegistry.register('1', 'background-color:red;');
```

### `registry.has(id)`

Returns true if the style id is already tracked by this registry

**Arguments:**

-   `id`: string - unique identifier for the provided style

**Example:**

```jsx
import registry from '@trousers/registry';

const clientRegistry = registry(
    document.createElement('head'),
    'i-love-trousers',
);

clientRegistry.has('1'); // Returns false!
clientRegistry.register('1', 'background-color:red;');
clientRegistry.has('1'); // Returns true!
```

### `registry.clear()`

Flushes all styles tracked by this registry

**Example:**

```jsx
import registry from '@trousers/registry';

const clientRegistry = registry(
    document.createElement('head'),
    'i-love-trousers',
);

clientRegistry.register('1', 'background-color:red;');
clientRegistry.clear(); // clears the style element attached to this registry
```
