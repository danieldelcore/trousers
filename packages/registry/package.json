{
    "name": "@trousers/registry",
    "version": "3.0.1",
    "main": "dist/index.js",
    "module": "dist/index.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {
        "@trousers/parser": "3.0.1",
        "stylis": "^3.5.4"
    },
    "devDependencies": {},
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/registry"
}
