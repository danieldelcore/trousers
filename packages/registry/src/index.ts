export { default as registry, Registry } from './registry';
