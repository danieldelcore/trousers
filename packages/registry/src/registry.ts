import stylis from 'stylis';

export interface Registry {
    register: (id: string, styles: string, isGlobal?: boolean) => void;
    has: (id: string) => boolean;
    clear: (isGlobal?: boolean) => void;
}

interface RegistryOptions {
    forceNewNode: boolean;
    appendBefore?: string;
}

const parse = (str: string) =>
    str
        .split('}')
        .filter(str => str !== '')
        .reduce<string[]>((accum, str, i) => {
            if (i > 0) {
                const prev = accum.length - 1;
                const openCount = (accum[prev].match(/{/g) || []).length;
                const closeCount = (accum[prev].match(/}/g) || []).length;

                if (openCount > closeCount) {
                    accum[prev] = accum[prev] + str + '}';
                    return accum;
                }
            }

            accum.push(str + '}');
            return accum;
        }, []);

const createStyleElement = (attributeId: string) => {
    const element = document.createElement<'style'>('style');
    element.setAttribute(attributeId, '');
    element.setAttribute('type', 'text/css');
    return element;
};

const getStyleElement = (targetElement: HTMLElement, attributeId: string) => {
    const element = targetElement.querySelector<HTMLStyleElement>(
        `style[${attributeId}]`,
    );

    return !!element ? element : createStyleElement(attributeId);
};

const registry = (
    parentElement: HTMLElement,
    attributeId: string,
    options: RegistryOptions = {
        forceNewNode: false,
        appendBefore: undefined,
    },
): Registry => {
    const styleElement = options.forceNewNode
        ? createStyleElement(attributeId)
        : getStyleElement(parentElement, attributeId);

    if (!options.appendBefore) {
        parentElement.appendChild(styleElement);
    } else {
        parentElement.insertBefore(
            styleElement,
            parentElement.querySelector<HTMLStyleElement>(
                `style[${options.appendBefore}]`,
            ),
        );
    }

    const clear = () => styleElement.remove();
    const has = (id: string): boolean =>
        styleElement.getAttribute(attributeId)!.includes(id);

    const register = (id: string, styles: string, isGlobal?: boolean) => {
        if (has(id)) return;

        const selector = !isGlobal ? id : ``;
        const processedStyles = stylis(selector, styles);

        if (process.env.NODE_ENV === 'production') {
            parse(processedStyles).forEach(styles => {
                // @ts-ignore
                styleElement.sheet.insertRule(
                    styles,
                    // @ts-ignore
                    isGlobal ? 0 : styleElement.sheet.cssRules.length,
                );
            });

            return;
        }

        const styleNode = document.createTextNode(`${processedStyles}\n`);
        const mountedStyles = styleElement.getAttribute(attributeId);

        styleElement.appendChild(styleNode);
        styleElement.setAttribute(attributeId, `${mountedStyles} ${id}`.trim());
    };

    return {
        has,
        clear,
        register,
    };
};

export default registry;
