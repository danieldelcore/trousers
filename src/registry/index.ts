export { default as Registry } from './registry';
export { default as ServerRegistry } from './server-registry';
