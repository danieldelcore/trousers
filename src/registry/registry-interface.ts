export default interface RegistryInterface {
    register: (id: string, styles: string, isGlobal: boolean) => void;
    has: (id: string) => boolean;
    clear: (isGlobal: boolean) => void;
}
