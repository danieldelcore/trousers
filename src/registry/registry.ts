import stylis from 'stylis';

import RegistryInterface from './registry-interface';

interface RegistryOptions {
    forceNewNode: boolean;
    appendBefore?: string;
}

class Registry implements RegistryInterface {
    private styleElement!: HTMLStyleElement;

    constructor(
        private parentElement: HTMLElement,
        private attributeId: string,
        private options: RegistryOptions = {
            forceNewNode: false,
            appendBefore: undefined,
        },
    ) {
        this.attributeId = attributeId;
        this.parentElement = parentElement;

        this.styleElement = this.options.forceNewNode
            ? this.createStyleElement()
            : this.getStyleElement();

        if (!this.options.appendBefore) {
            this.parentElement.appendChild(this.styleElement);
        } else {
            this.parentElement.insertBefore(
                this.styleElement,
                this.parentElement.querySelector<HTMLStyleElement>(
                    `style[${this.options.appendBefore}]`,
                ),
            );
        }
    }

    register(id: string, styles: string, isGlobal?: boolean) {
        if (this.has(id)) return;

        const selector = !isGlobal ? id : ``;
        const processedStyles = stylis(selector, styles);
        const styleNode = document.createTextNode(`${processedStyles}\n`);
        const mountedStyles = this.styleElement.getAttribute(this.attributeId);

        this.styleElement.appendChild(styleNode);
        this.styleElement.setAttribute(
            this.attributeId,
            `${mountedStyles} ${id}`.trim(),
        );
    }

    has(id: string): boolean {
        const mountedStyles = this.styleElement.getAttribute(this.attributeId);

        return mountedStyles!.includes(id);
    }

    clear() {
        this.styleElement.remove();
    }

    private getStyleElement(): HTMLStyleElement {
        const element = this.parentElement.querySelector<HTMLStyleElement>(
            `style[${this.attributeId}]`,
        );

        return !!element ? element : this.createStyleElement();
    }

    private createStyleElement(): HTMLStyleElement {
        const element = document.createElement<'style'>('style');
        element.setAttribute(this.attributeId, '');
        element.setAttribute('type', 'text/css');
        element.type = 'text/css';

        return element;
    }
}

export default Registry;
