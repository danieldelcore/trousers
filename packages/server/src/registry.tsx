import React, { ReactNode, ReactChild } from 'react';
import stylis from 'stylis';

import { STYLE_ID, GLOBAL_STYLE_ID } from '@trousers/utils';
import { Registry } from '@trousers/registry';

export type ServerRegistry = Registry & { get: () => ReactNode };

const serverRegistry = (): ServerRegistry => {
    const globalMap = new Map<string, string>();
    const styleMap = new Map<string, string>();
    const has = (id: string) => styleMap.has(id);

    const clear = () => {};
    const register = (id: string, style: string, isGlobal?: boolean) => {
        if (has(id)) return;

        const selector = !isGlobal ? id : ``;
        const processedStyles: string = stylis(selector, style);

        if (!isGlobal) {
            styleMap.set(id, processedStyles);
        } else {
            globalMap.set(id, processedStyles);
        }
    };

    const renderStyles = () => {
        const ids: string[] = [];
        const styles: string[] = [];

        styleMap.forEach((value, key) => {
            ids.push(key);
            styles.push(value);
        });

        return (
            <style
                {...{
                    [STYLE_ID]: ids.toString(),
                }}
            >
                {styles.join('\n')}
            </style>
        );
    };

    const renderGlobals = () => {
        const styles: ReactChild[] = [];

        globalMap.forEach((value, key) => {
            styles.push(
                <style
                    {...{
                        [GLOBAL_STYLE_ID]: key,
                    }}
                >
                    {value}
                </style>,
            );
        });

        return styles;
    };

    const get = () => [...renderGlobals(), renderStyles()];

    return {
        get,
        has,
        clear,
        register,
    };
};

export default serverRegistry;
