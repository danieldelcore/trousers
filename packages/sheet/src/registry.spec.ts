import registry, { Registry } from './registry';

describe('Registry', () => {
    let clientRegistry: Registry;
    let element: HTMLHeadElement;
    let attributeId: string;

    beforeEach(() => {
        element = document.createElement('head');
        attributeId = 'data-testing-trousers';

        clientRegistry = registry(element, attributeId);
    });

    it('registers a new style', () => {
        const key = '1';
        const style = 'background-color:red;';

        clientRegistry.register(key, style);

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('registers multiple styles', () => {
        clientRegistry.register('1', 'background-color:red;');
        clientRegistry.register('2', 'background-color:blue;');

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('registers a global style', () => {
        const key = '1';
        const style = 'background-color:red;';

        clientRegistry.register(key, style, true);

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('registers a global, with preexisting style tag', () => {
        clientRegistry.register('1', 'background-color:red;');
        clientRegistry.register('2', 'background-color:blue;', true);

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('registering an existing key does not update the style', () => {
        const key = '1';
        const style = 'background-color:red;';
        const styleBlue = 'background-color:blue;';

        clientRegistry.register(key, style);
        clientRegistry.register(key, styleBlue);

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('detects an existing style', () => {
        const key = '1';
        const style = 'background-color:red;';

        clientRegistry.register(key, style);

        expect(clientRegistry.has('1')).toBe(true);
    });

    it('detects a non existing style', () => {
        expect(clientRegistry.has('123')).toBe(false);
    });

    it('registries can mount their own style tags', () => {
        const secondaryRegistry = registry(element, 'data-testing-secondary', {
            forceNewNode: true,
        });

        clientRegistry.register('1', 'background-color:red;');
        secondaryRegistry.register('2', 'background-color:blue;');

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('registries can hoist style tags to the top', () => {
        const secondaryRegistry = registry(element, 'data-testing-secondary', {
            forceNewNode: true,
            appendBefore: attributeId,
        });

        clientRegistry.register('1', 'background-color:red;');
        secondaryRegistry.register('2', 'background-color:blue;', true);

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('registries will append style node if append before selector returns null', () => {
        const secondaryRegistry = registry(element, 'data-testing-secondary', {
            forceNewNode: true,
            appendBefore: 'data-unknown-element',
        });

        secondaryRegistry.register('1', 'background-color:blue;');

        expect(element.innerHTML).toMatchSnapshot();
    });

    it('clears multiple mounted style tags', () => {
        const secondaryRegistry = registry(element, 'data-testing-secondary', {
            forceNewNode: true,
        });

        clientRegistry.register('1', 'background-color:red;');
        secondaryRegistry.register('2', 'background-color:blue;');

        expect(element.childElementCount).toEqual(2);

        secondaryRegistry.clear();

        expect(element.childElementCount).toEqual(1);

        clientRegistry.clear();

        expect(element.childElementCount).toEqual(0);
    });
});
