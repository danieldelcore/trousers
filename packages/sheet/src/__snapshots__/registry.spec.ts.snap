// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry registering an existing key does not update the style 1`] = `
<style data-testing-trousers="1" type="text/css">
  1{background-color:red;}
</style>
`;

exports[`Registry registers a global style 1`] = `
<style data-testing-trousers="1" type="text/css">
  {background-color:red;}
</style>
`;

exports[`Registry registers a global, with preexisting style tag 1`] = `
<style data-testing-trousers="1 2" type="text/css">
  1{background-color:red;}
  {background-color:blue;}
</style>
`;

exports[`Registry registers a new style 1`] = `
<style data-testing-trousers="1" type="text/css">
  1{background-color:red;}
</style>
`;

exports[`Registry registers multiple styles 1`] = `
<style data-testing-trousers="1 2" type="text/css">
  1{background-color:red;}
  2{background-color:blue;}
</style>
`;

exports[`Registry registries can hoist style tags to the top 1`] = `
<style data-testing-secondary="2" type="text/css">
  {background-color:blue;}
</style>
<style data-testing-trousers="1" type="text/css">
  1{background-color:red;}
</style>
`;

exports[`Registry registries can mount their own style tags 1`] = `
<style data-testing-trousers="1" type="text/css">
  1{background-color:red;}
</style>
<style data-testing-secondary="2" type="text/css">
  2{background-color:blue;}
</style>
`;

exports[`Registry registries will append style node if append before selector returns null 1`] = `
<style data-testing-trousers="" type="text/css"></style>
<style data-testing-secondary="1" type="text/css">
  1{background-color:blue;}
</style>
`;
